### PR TITLE
feat: add full screen ad hooks

### PR DESCRIPTION
**Pull request by wjaykim**
### Description

This PR add supports for hook-style ad state management which can be used for full screen ads(AppOpen, Interstitial, Rewarded). This feature is initially implemented in another admob library, [@react-native-admob/admob](https://github.com/react-native-admob/admob).
I'm happy to join forces on this single package which can be maintained better😁.

### Release Summary

add full screen ad hooks

### Checklist

- I read the [Contributor Guide](https://github.com/invertase/react-native-google-mobile-ads/blob/main/CONTRIBUTING.md)
  and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `__tests__e2e__`
  - [ ] `jest` tests added or updated in `__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [ ] No

### Test Plan

---

Think `react-native-google-mobile-ads` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Comment by mikehardy** [comment]
@wjaykim welcome welcome 👋 😄  I really excited personally to (hopefully) have everyone join forces in one library we can all tag-team on so as maintainers we all have just a little to do instead of everyone having to do everything. I'll queue this for review and hopefully the vision of one great admob package and a reasonable work load for everyone can become a reality :)

File: docs/displaying-ads-hook.mdx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Hooks
+
+The AdMob package provides hooks to help you to display ads in functional component with tiny code. The supported ad formats are full-screen-type-ads: AppOpen, Interstitial, Rewarded.
```
Comment (dylancom) [inline docs/displaying-ads-hook.mdx:3]:
```suggestion
The AdMob package provides hooks to help you to display ads in a functional component with tiny code. The supported ad formats are full-screen ads: App open, Interstitial & Rewarded.
```

File: docs/displaying-ads-hook.mdx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+## Creating ad
```
Comment (dylancom) [inline docs/displaying-ads-hook.mdx:5]:
```suggestion
## Load an ad
```

File: docs/displaying-ads-hook.mdx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+> `adUnitid` parameter is also can be used to manage creation and destruction of ad instance.
+> If `adUnitid` is set or changed, new ad instance will be created and previous ad instance will be destroyed if exists.
+> If `adUnitid` is set to `null`, no ad instance will be created and previous ad instance will be destroyed if exists.
+
+The second argument is additional optional request options object to be sent whilst loading an advert, such as keywords & location.
+Setting additional request options helps AdMob choose better tailored ads from the network. View the [`RequestOptions`](/reference/admob/requestoptions)
+documentation to view the full range of options available.
+
+## Displaying ad
```
Comment (dylancom) [inline docs/displaying-ads-hook.mdx:33]:
```suggestion
## Show the ad
```

File: docs/displaying-ads-hook.mdx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+export default function App() {
+  const interstitialAd = useInterstitialAd(TestIds.Interstitial, {
+    requestNonPersonalizedAdsOnly: true,
+  });
+
+  return <View>{/* ... */}</View>;
+}
+```
+
+> `adUnitid` parameter is also can be used to manage creation and destruction of ad instance.
```
Comment (dylancom) [inline docs/displaying-ads-hook.mdx:25]:
```suggestion
> The `adUnitid` parameter can also be used to manage creation and destruction of an ad instance.
```

File: docs/displaying-ads-hook.mdx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+> If `adUnitid` is set or changed, new ad instance will be created and previous ad instance will be destroyed if exists.
+> If `adUnitid` is set to `null`, no ad instance will be created and previous ad instance will be destroyed if exists.
+
+The second argument is additional optional request options object to be sent whilst loading an advert, such as keywords & location.
```
Comment (dylancom) [inline docs/displaying-ads-hook.mdx:29]:
```suggestion
The second argument is an additional optional request options object to be sent whilst loading an advert, such as keywords & location.
```

File: src/hooks/useAppOpenAd.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ */
+export function useAppOpenAd(
+  adUnitId: string | null,
+  requestOptions: RequestOptions = {},
+): Omit<AdHookReturns, 'adReward'> {
+  const [appOpenAd, setAppOpenAd] = useState<AppOpenAd | null>(null);
+
+  useDeepCompareEffect(() => {
+    setAppOpenAd(() => {
+      //prevAd?.destroy();
```
Comment (dylancom) [inline src/hooks/useAppOpenAd.ts:41]:
```suggestion
```

File: src/hooks/useInterstitialAd.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ */
+export function useInterstitialAd(
+  adUnitId: string | null,
+  requestOptions: RequestOptions = {},
+): Omit<AdHookReturns, 'adReward'> {
+  const [interstitialAd, setInterstitialAd] = useState<InterstitialAd | null>(null);
+
+  useDeepCompareEffect(() => {
+    setInterstitialAd(() => {
+      //prevAd?.destroy();
```
Comment (dylancom) [inline src/hooks/useInterstitialAd.ts:41]:
```suggestion
```

File: src/hooks/useRewardedAd.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ */
+export function useRewardedAd(
+  adUnitId: string | null,
+  requestOptions: RequestOptions = {},
+): Omit<AdHookReturns, 'adReward'> {
+  const [rewardedAd, setRewardedAd] = useState<RewardedAd | null>(null);
+
+  useDeepCompareEffect(() => {
+    setRewardedAd(() => {
+      //prevAd?.destroy();
```
Comment (dylancom) [inline src/hooks/useRewardedAd.ts:41]:
```suggestion
```

**Review by dylancom — CHANGES_REQUESTED** [review]

**Review by mikehardy — APPROVED** [review]
Looks great to me, and @dylancom just had language comments which I went ahead and just committed, since it seemed best to do a squash merge anyway since it's one feature all at once effectively

I'll pull locally, check the example and assuming it looks good merge this

**Review by dylancom — APPROVED** [review]

**Comment by mikehardy** [comment]
I tested it locally and everything works - I actually extended the test app a bit to test all 3 of the hooks and they're all good.
I would have included that as a new commit on this PR but I think I saw an error that testing exposed, so I'll do it as another PR.

So this is good to go --> @wjaykim :trophy: :bow:

**Pull request merged**